File: backend/app/schemas/coding_interview.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

Difficulty = Literal["easy", "medium", "hard"]


class CodingExample(BaseModel):
    input: str
    output: str
    explanation: str = ""


class CodingTestCase(BaseModel):
    input: str
    expected_output: str


class ExpectedComplexity(BaseModel):
    time: str = ""
    space: str = ""


class CodingQuestion(BaseModel):
    model_config = ConfigDict(extra="ignore")

    question_number: int
    title: str
    problem_statement: str
    difficulty: Difficulty = "medium"
    topics: list[str] = Field(default_factory=list)
    constraints: list[str] = Field(default_factory=list)
    examples: list[CodingExample] = Field(default_factory=list)
    function_signature: str = ""
    starter_code: dict[str, str] = Field(default_factory=dict)
    visible_test_cases: list[CodingTestCase] = Field(default_factory=list)
    hidden_test_cases: list[CodingTestCase] = Field(default_factory=list)
    expected_complexity: ExpectedComplexity = Field(default_factory=ExpectedComplexity)


class CodingQuestionPublic(CodingQuestion):
    hidden_test_cases: list[CodingTestCase] = Field(default_factory=list, exclude=True)


class CodingInterviewStartRequest(BaseModel):
    resume_id: str | None = None
    role: str = Field(min_length=2, max_length=120)
    experience_level: str = Field(default="Fresher", min_length=2, max_length=80)
    difficulty: Difficulty = "medium"
    topics: list[str] = Field(default_factory=list, max_length=10)
    language: str = "cpp"
    number_of_questions: int = Field(default=3, ge=1, le=5)


class CodingInterviewSession(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str | None = None
    user_id: str
    resume_id: str | None = None
    role: str
    experience_level: str
    difficulty: Difficulty
    language: str
    questions: list[CodingQuestion]
    completed: bool = False
    final_score: float = 0.0
    created_at: datetime = Field(default_factory=datetime.utcnow)
    completed_at: datetime | None = None


class CodingInterviewResponse(BaseModel):
    session_id: str
    role: str
    difficulty: Difficulty
    language: str
    questions: list[CodingQuestionPublic]
    completed: bool
    final_score: float


class CodingSubmitRequest(BaseModel):
    question_number: int = Field(ge=1)
    language: str
    source_code: str = Field(min_length=1)
    explanation: str = ""


class TestExecutionResult(BaseModel):
    input: str
    expected_output: str
    actual_output: str = ""
    passed: bool
    stderr: str = ""
    execution_time: float | None = None


class CodingSubmission(BaseModel):
    id: str | None = None
    session_id: str
    user_id: str
    question_number: int
    language: str
    source_code: str
    explanation: str = ""
    test_results: list[TestExecutionResult] = Field(default_factory=list)
    tests_passed: int = 0
    tests_total: int = 0
    correctness_score: float = 0.0
    quality_score: float = 0.0
    complexity_score: float = 0.0
    explanation_score: float = 0.0
    overall_score: float = 0.0
    feedback: list[str] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=datetime.utcnow)


class CodingSubmitResponse(BaseModel):
    submission: CodingSubmission
    next_question: CodingQuestionPublic | None = None


class CodingHistoryResponse(BaseModel):
    sessions: list[CodingInterviewResponse]
    page: int
    limit: int
    total: int


class SupportedLanguagesResponse(BaseModel):
    provider: str
    languages: list[str]

File: frontend/src/types/codingInterview.ts
/**
 * Mirrors backend/app/schemas/coding_interview.py exactly.
 *
 * IMPORTANT (confirmed by reading app/api/v1/coding_interview.py):
 *   - POST /coding-interviews/start           returns the full question set
 *     (hidden_test_cases excluded via CodingQuestionPublic)
 *   - GET  /coding-interviews/{id}             also returns the full
 *     question set, so a session IS resumable after a refresh (unlike the
 *     text-interview module).
 *   - POST /coding-interviews/{id}/run         runs only visible test
 *     cases, no persistence, no scoring.
 *   - POST /coding-interviews/{id}/submit      runs all test cases,
 *     scores, persists, returns the next question.
 *   - POST /coding-interviews/{id}/complete    finalizes the session.
 */

export type CodingDifficulty = "easy" | "medium" | "hard";

export interface CodingExample {
  input: string;
  output: string;
  explanation: string;
}

export interface CodingTestCase {
  input: string;
  expected_output: string;
}

export interface ExpectedComplexity {
  time: string;
  space: string;
}

export interface CodingQuestion {
  question_number: number;
  title: string;
  problem_statement: string;
  difficulty: CodingDifficulty;
  topics: string[];
  constraints: string[];
  examples: CodingExample[];
  function_signature: string;
  starter_code: Record<string, string>;
  visible_test_cases: CodingTestCase[];
  expected_complexity: ExpectedComplexity;
}

export interface CodingInterviewStartPayload {
  resume_id?: string | null;
  role: string;
  experience_level: string;
  difficulty: CodingDifficulty;
  topics: string[];
  language: string;
  number_of_questions: number;
}

export interface CodingInterviewResponse {
  session_id: string;
  role: string;
  difficulty: CodingDifficulty;
  language: string;
  questions: CodingQuestion[];
  completed: boolean;
  final_score: number;
}

export interface CodingRunPayload {
  question_number: number;
  language: string;
  source_code: string;
}

export interface TestExecutionResult {
  input: string;
  expected_output: string;
  actual_output: string;
  passed: boolean;
  stderr: string;
  execution_time: number | null;
}

export interface CodingRunResponse {
  test_results: TestExecutionResult[];
  tests_passed: number;
  tests_total: number;
}

export interface CodingSubmitPayload {
  question_number: number;
  language: string;
  source_code: string;
  explanation?: string;
}

export interface CodingSubmission {
  id: string;
  session_id: string;
  question_number: number;
  language: string;
  source_code: string;
  explanation: string;
  test_results: TestExecutionResult[];
  tests_passed: number;
  tests_total: number;
  correctness_score: number;
  quality_score: number;
  complexity_score: number;
  explanation_score: number;
  overall_score: number;
  feedback: string[];
  created_at: string;
}

export interface CodingSubmitResponse {
  submission: CodingSubmission;
  next_question: CodingQuestion | null;
}

export interface CodingHistoryResponse {
  sessions: CodingInterviewResponse[];
  page: number;
  limit: number;
  total: number;
}

export interface SupportedLanguagesResponse {
  provider: string;
  languages: string[];
}

export const CODING_DIFFICULTIES: CodingDifficulty[] = ["easy", "medium", "hard"];
